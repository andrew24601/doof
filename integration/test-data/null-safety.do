// Test null safety operators (optional chaining and nullish coalescing)

// Nullable types and null coalescing operator


// Optional chaining with objects



// Optional chaining with methods
    
    



// Non-null assertion operator (if supported)


class Address {
    street: string | null;
    city: string | null;
    
}

    class Person {
        name: string;
        address: Address | null;
    }

    class Calculator {
        value: int = 0;
        add(x: int): Calculator {
            this.value += x;
            return this;
        }
        getResult(): int {
            return this.value;
        }
    }


function main(): int {
    let result = "";
    let nullableString: string | null = null;
    let defaultValue = nullableString ?? "default";
    result = result + defaultValue + "|";
    nullableString = "actual";
    let actualValue = nullableString ?? "default";
    result = result + actualValue + "|";
    let person: Person | null = null;
    let cityName = person?.address?.city ?? "unknown";
    result = result + cityName + "|";
    person = {
        name: "Alice",
        address: {
            street: "123 Main St",
            city: "Anytown"
        }
    };
    cityName = person?.address?.city ?? "unknown";
    result = result + cityName + "|";
    let calc: Calculator | null = new Calculator();
    let chainResult = calc?.add(5)?.add(3)?.getResult() ?? 0;
    result = result + chainResult + "|";
    calc = null;
    chainResult = calc?.add(5)?.add(3)?.getResult() ?? 999;
    result = result + chainResult + "|";
    let definitelyString: string | null = "hello";
    let assertedLength = definitelyString!.length;
    result = result + assertedLength;
    println(result);
    return 0;
}