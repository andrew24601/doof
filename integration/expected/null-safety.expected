default|actual|unknown|Anytown|8|999|5